test/patterns_as_and_drop/test.ml

```diff
@@ -0,0 +1,69 @@
+open Ppxlib
+open Format
+
+let loc = Location.none
+
+let ast = [%expr List.length xs = 0]
+
+(* Simple demo without [as__] and [drop] *)
+let () =
+  let pat =
+    let open Ast_pattern in
+    let length () =
+      pexp_apply
+        (pexp_ident (ldot (lident (string "List")) (string "length")))
+        ((nolabel ** __) ^:: nil)
+    in
+    let zero () = pexp_constant (pconst_integer (string "0") none) in
+
+    pexp_apply
+      (pexp_ident (lident (string "=")))
+      ((nolabel ** length ()) ^:: (nolabel ** zero ()) ^:: nil)
+  in
+
+  Ast_pattern.parse pat loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun _ -> print_endline "success")
+
+let () =
+  let pat =
+    let open Ast_pattern in
+    let length () =
+      as__
+        (pexp_apply
+           (pexp_ident (ldot (lident (string "List")) (string "length")))
+           ((nolabel ** drop) ^:: nil))
+    in
+    let zero () = as__ (pexp_constant (pconst_integer (string "0") none)) in
+
+    pexp_apply
+      (pexp_ident (lident (string "=")))
+      ((nolabel ** length ()) ^:: (nolabel ** zero ()) ^:: nil)
+  in
+
+  Ast_pattern.parse pat loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun _l _r ->
+      (* Format.printf "'%a' and '%a'\n%!" Pprintast.expr l Pprintast.expr r; *)
+      (* Error: The module Pprintast is an alias for module Astlib.Pprintast, which is missing *)
+      Format.printf "Success but '??' and '??'\n%!")
+
+(** Pitfall. If we forget unit argument and will use [as__], the success case
+    will be fired before the error case. *)
+let () =
+  let loc = Location.none in
+  let ast = [%expr 1] in
+
+  let pat () =
+    let open Ast_pattern in
+    as__ (pexp_constant (pconst_integer (string "0") none))
+  in
+  printf "Wrong!\n%!";
+  Ast_pattern.parse (pat ()) loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun _zero_expr -> Format.printf "Success and error after that\n%!");
+  printf "Correct!\n%!";
+  Ast_pattern.parse (pat ()) loc ast
+    ~on_error:(fun () () -> print_endline "error")
+    (fun _zero_expr () -> Format.printf "Success and error after that\n%!")
+    ()
```

Comment:
Am I understanding correctly that here the context is a function from `unit` to `string`?

Reply:
I don't understand a question. Ast_pattern0 defines type `context ` which is isomorphic to integer, so it is not a function.

Reply:
You're right, I need to explain myself better since in ppxlib there are lots of different concepts that we call context. What I was referring to is the type parameter `'c` in `('a, 'b, 'c) Ast_pattern.t` (i.e. the return type of the continuation `'b`). Usually, that type parameter stands for the context inside the AST in which you want to inject the derived/expanded code.

Reply:
Context is `unit -> unit` because both `() -> print_endline "error"` and `fun () -> Format.printf "Success and error after that\n%!"` have this type